sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -687,13 +687,48 @@ def get_cmis_media_lanes_mask(self, api, appl, lport, subport):
 
         return media_lanes_mask
 
-    def is_appl_reconfigure_required(self, api, app_new):
+    def set_dp_decommision_pending(self, lport, decomm_pending):
         """
-	   Reset app code if non default app code needs to configured 
+        Set decommission pending flag for the datapath of the given logical port
+        and force CMIS reinit if decomm_pending is False.
+
+        Args:
+            lport:
+                String, logical port name
+            decomm_pending:
+                Boolean, True, if decommission pending is set, False otherwise
+                         False, it means the datapath is ready to be reinitialized.
+        """
+        physical_port = self.port_dict[lport]['index']
+        for logical_port, pdata in self.port_dict.items():
+            if pdata['index'] == physical_port:
+                pdata['is_decomm_pending'] = decomm_pending
+                if not decomm_pending:
+                    self.force_cmis_reinit(logical_port, 0)
+
+    def is_decommission_required(self, lport, api):
         """
+        Reset app code if non default app code needs to configured
+
+        Args:
+            api:
+                XcvrApi object
+            lport:
+                String, logical port name
+
+        Returns:
+            True, if decommission is required
+            False, if decommission is done or not required
+        """
+        if 'is_decomm_pending' in self.port_dict[lport]:
+            return self.port_dict[lport]['is_decomm_pending']
+
         for lane in range(self.CMIS_MAX_HOST_LANES):
-            app_cur = api.get_application(lane)
-            if app_cur != 0 and app_cur != app_new:
+            app = api.get_application(lane) & 0xF
+            if app != 0:
+                #decommission all lanes/datapaths to default AppSel=0
+                self.log_notice(f"{lport}: Decommissioning: set all lanes/datapaths to default AppSel=0")
+                self.set_dp_decommision_pending(lport, True)
```

Comment:
@AnoopKamath for readability can you split this into two function where the cmis reinit is done separately and pending is set separately.

Reply:
Thanks @prgeor, taken care